### PR TITLE
Added demandResponse.go file to implement DR listener function

### PR DESCRIPTION
**Pull request by john-b-yang**

### CODE REVIEW

File: driver/pelican/demandResponse.go
Comment (gtfierro) [inline driver/pelican/demandResponse.go]: Can we make this function a comment? Or put it into a separate `test` folder? I'd like the functions in this package to be for the operation of the thermostat and manipulation of its APIs

File: driver/pelican/demandResponse.go
Comment (gtfierro) [inline driver/pelican/demandResponse.go]: Can we implement this as a set of constants?

```go
type DR_EVENT_STATUS int
const (
  DR_EVENT_UNKNOWN = -1
  DR_EVENT_NOT_CONFIGURED = iota
  DR_EVENT_UNUSABLE
  DR_EVENT_INACTIVE
// etc...
)
```

Then we can have another function that translates a string to a `DR_EVENT_STATUS`.

File: driver/pelican/demandResponse.go
Comment (gtfierro) [inline driver/pelican/demandResponse.go]: Let's implement an enum construction here too. We'll be able to use these enum definitions in other DR-related drivers and applications

**Review by gtfierro — CHANGES_REQUESTED** [review]
Good start! There's a few small things that I'd like changed (below)

Are you going to implement the XBOS DR interface on top of this, or will that be a forthcoming pull request?

File: driver/pelican/demandResponse.go
Comment (jhkolb) [inline driver/pelican/demandResponse.go]: I believe the convention in Go is to use camel case, not underscores, so you'll need to change several of the fields in this struct.

File: driver/pelican/demandResponse.go
Comment (jhkolb) [inline driver/pelican/demandResponse.go]: Let's make this error message more specific, we're trying to retrieve information about a DR event.

File: driver/pelican/demandResponse.go
Comment (jhkolb) [inline driver/pelican/demandResponse.go]: Why is the `&` operator used here?

File: driver/pelican/demandResponse.go
Comment (jhkolb) [inline driver/pelican/demandResponse.go]: And let's make this error message clearer too.

File: driver/pelican/demandResponse.go
Comment (jhkolb) [inline driver/pelican/demandResponse.go]: Am I confused, or are the two conditions you're checking for redundant? I guess I could imagine some weirdness with non-printable characters, but I would think we only need to check for a length of 0.

File: driver/pelican/demandResponse.go
Comment (jhkolb) [inline driver/pelican/demandResponse.go]: There's no casting going on here, and there's no int64 value yet either.  This will produce a `time.Time` instance with the specified time zone.

File: driver/pelican/demandResponse.go
Comment (jhkolb) [inline driver/pelican/demandResponse.go]: Hm I don't think I've seen `XMLName` used in the rest of the Pelican code. Why do we need it here? If we don't, we probably want to keep things consistent.

**Review by jhkolb — CHANGES_REQUESTED** [review]
I have a few other suggestions for things to change as well. Did you forget to commit your changes to the `main.go` file, which will actually perform the Bosswave operations and, as Gabe mentioned, complete the implementation of the XBOS interface?

**Comment by john-b-yang** [comment]
@jhkolb @gtfierro Thanks for the feedback! I've updated the code to reflect the requested edits. Most significantly, I introduced two sets of an enum and a helper method to better define a DR event's Status and Type. I also removed the test method I originally wrote and folded the logic for publishing the DR status into 'main.go'.

File: driver/pelican/demandResponse.go
Comment (gtfierro) [inline driver/pelican/demandResponse.go]: `direct-response` should be changed to `demand response` here and elsewhere

File: driver/pelican/demandResponse.go
Comment (gtfierro) [inline driver/pelican/demandResponse.go:40]: Can we annotate this struct so that the msgpack serialization is compliant with the [XBOS DR Interface](https://github.com/SoftwareDefinedBuildings/XBOSDocs/blob/master/driver_xbos_demand_response.md)?

Should be able to do this like

```go
type ADREvent struct {
    EventEnd int64 `msgpack:"event_end"`
    // etc
```

**Review by gtfierro — APPROVED** [review]

**Comment by gtfierro** [comment]
Thanks @john-b-yang! This looks 95% good to me. The one last change, which I should have caught earlier (sorry!), is the `i.xbos.demand_response` interface is separate from the `i.xbos.thermostat` interface.

We should have a similar line to https://github.com/john-b-yang/bw2-contrib/blob/ea9f2fd33db9aac9a4676e6e824181a4021b4a69/driver/pelican/main.go#L63 that instantiates a map of `i.xbos.demand_response` interfaces, and then [here](https://github.com/john-b-yang/bw2-contrib/blob/ea9f2fd33db9aac9a4676e6e824181a4021b4a69/driver/pelican/main.go#L168) you should make sure you're publishing from the demand response interface instead of the thermostat interface. The name of the signal should be `info`, not `dr` as well. This is all to just make it compliant with the `i.xbos.demand_response` interface description.

When this is behaving correctly, we should see URIs like 

```
Message from <namespace>/pelican/s.pe  
lican/SouthEastCorner/i.xbos.demand_response/signal/info:  
PO 2.1.1.9 len 62 (msgpack) contents:  
{  
  "dr_status": 3,  
  "event_end": 0,  
  "event_start": 0,  
  "event_type": -1,  
  "time": 1525105880378267371  
}
```

File: driver/pelican/main.go
Comment (jhkolb) [inline driver/pelican/main.go]: I think you want `drErr` rather than `err` in this print.

**Comment by gtfierro** [comment]
Thanks @john-b-yang ; does the Pelican DR interface expose a timestamp that we're going to have to deal with, like we do with the current thermostat reading?

**Comment by jhkolb** [comment]
I don't think the DR API endpoint exposes a timestamp.

Also, do we want to put the publish loops for the two interfaces in separate goroutines? For example, what if we want to poll the thermostat and DR endpoints at different frequencies? This could be overcomplicating things though, so I'm not sure it's really necessary.

**Comment by gtfierro** [comment]
Oh, good idea! We'll probably only need to poll the DR interface every hour or so? The pelican thermostat polling is already inside a goroutine so I think that should be a straightforward change. The poll rate for the demand response interface should be another config option.

**Comment by john-b-yang** [comment]
@gtfierro @jhkolb I created a separate go routine and poll interval for the demand response polling!

File: driver/pelican/main.go
Comment (gtfierro) [inline driver/pelican/main.go]: Probably put `drErr` here instead of `err`

**Comment by gtfierro** [comment]
I just had one minor fix and then I'm ready to merge this predicated on @jhkolb approval

File: driver/pelican/demandResponse.go
Comment (jhkolb) [inline driver/pelican/demandResponse.go]: Strictly speaking, -1 is not a valid value for either the DR event status or type in the current specification of the XBOS interface.

Do we want to remove those values here or update the interface spec to incorporate them?

Comment (gtfierro) [reply driver/pelican/demandResponse.go]: I like the idea of being able to indicate when something is not understood, but to remain consistent with the other interfaces maybe we want to just drop it when we don't know what's going on (this is the behavior other drivers exhibit). I'm ok with removing these and just printing out an error when we do get an unexpected indicator

Comment (john-b-yang) [reply driver/pelican/demandResponse.go]: Sounds good! I'll make sure to return an fmt.Error as opposed to assigning an actual error. In that case, I will probably just removed the "unknown" and "unrecognized" cases for the event status and type structs.

File: driver/pelican/main.go
Comment (john-b-yang) [reply driver/pelican/main.go]: Ahhh I keep making these mistakes, I'll be more thorough in the future, sorry about that haha.

**Comment by jhkolb** [comment]
This looks ready to me!

**Review by jhkolb — APPROVED** [review]

**Comment by jhkolb** [comment]
I'll leave it to @gtfierro to check that the recent fixes are sufficient and then merge.

File: driver/pelican/main.go
Comment (gtfierro) [inline driver/pelican/main.go:175]: A problem I noticed when testing is that if we throw an error when parsing the DR event, then the driver exits. This means that we won't be able to run Pelican drivers unless there's a DR event. Maybe we just log the error here and don't cause the DR event to crash the driver? What do folks think?

**Comment by jhkolb** [comment]
Ah, now I see why we had the previous design.

I agree, let's log the fact that we didn't retrieve a valid DR event, but keep the driver running.

File: driver/pelican/main.go
Comment (john-b-yang) [reply driver/pelican/main.go:175]: @gtfierro @jhkolb That sounds good to me. I can just leave the print statement and remove the "done <- true" statement so driver get request loop does not stop running?

Comment (gtfierro) [reply driver/pelican/main.go:175]: Yes that should work. Just make sure that the driver doesn't then try to publish an empty object. Maybe do the print, then sleep for the poll duration, and then `continue`

Comment (gtfierro) [reply driver/pelican/main.go:175]: Or alternatively set the drStatus to `nil`

**Comment by gtfierro** [comment]
Putting this on hold until https://github.com/SoftwareDefinedBuildings/bw2-contrib/pull/25 is fixed

**Comment by gtfierro** [comment]
I fixed up the merge and adjusted the behavior in https://github.com/SoftwareDefinedBuildings/bw2-contrib/pull/26

**Pull request closed without merging**
